### PR TITLE
 `TryGetSAMGeometries` Deep Optimization & Refactoring

### PR DESCRIPTION
**Pull request by ZiolkowskiJakub**
# Summary of `TryGetSAMGeometries` Deep Optimization, Refactoring & Test Suite

### 1. Open Brep / Non-Solid Brep -> `Shell` Target Combination
- **Addressed Review Item #1**: When requesting `T = Shell` and converting an open (non-solid) Brep or `IEnumerable<Face3D>` collection, all faces are now grouped into a single unified `Shell(face3Ds)` rather than producing degenerate single-face shells.

### 2. Edge Extraction Deduplication & Maintainability
- **Addressed Review Item #3**: Extracted shared edge retrieval logic across `Face3D`, `IFace3DObject`, `Shell`, and `IClosedPlanar3D` into a single helper method `GetEdges(object value)`.

### 3. Target Type Ordering & Interface Shadowing Fix
- Concrete requested target types (`Face3D`, `Polyline3D`, `Polygon3D`, `Segment3D`, `Shell`, `Mesh3D`, `Plane`, `Point3D`) execute **BEFORE** interface target types (`ISegmentable3D`, `IClosedPlanar3D`, `IFace3DObject`).
- Eliminated type shadowing bug where `typeof(ISegmentable3D).IsAssignableFrom(targetType)` matched concrete types first.
- Strict type checking (`is T targetItem`) is enforced on every item before adding to result list to prevent runtime `InvalidCastException`.

### 4. Dynamic Dispatch Elimination & Unwrapping Pipeline
- Replaced slow `as dynamic` DLR runtime binding with strongly-typed C# pattern matching.
- Implemented strongly-typed Grasshopper Goo and `global::Rhino.Geometry.*` unwrapping for `GH_Point`, `GH_Plane`, `GH_Vector`, `GH_Line`, `GH_Rectangle`, `GH_Circle`, `GH_Curve`, `GH_Surface`, `GH_Brep`, `GH_Mesh`, `GH_Extrusion`.
- Documented `IEnumerable` conversion assumption for collection inputs (`List<ISAMGeometry>`, `GH_Structure`, etc.).

### 5. Comprehensive Unit Test Suite Added
- **Addressed Review Item #2**: Created `TryGetSAMGeometriesTests.cs` covering:
  - `Face3D` extraction from direct `Face3D`, `GooSAMGeometry`, `Shell`, and closed `Polyline3D`.
  - `Polyline3D` extraction from `Face3D` (verifying zero `InvalidCastException`).
  - `Segment3D` extraction from `Face3D` and `Polyline3D`.
  - `Shell` extraction from multi-face `Face3D` collections.
  - `Plane` and `Point3D` extraction.
  - Collection processing with leading invalid/unconvertible elements (verifying no premature abort).
  - Null/empty guard validation.

## Verification
- **Build**: `dotnet build SAM.sln` completed with **0 Errors**.
- **Tests**: `dotnet test` executed cleanly (**24 Passed, 0 Failed** in `SAM.Core.Grasshopper.Tests.dll`).

### CODE REVIEW

File: Grasshopper/SAM.Geometry.Grasshopper/Query/TryGetSAMGeometries.cs [hunk 1]
```diff
@@ -6,314 +6,603 @@
...
+                        }
                     }
+                    return;
                 }
             }
 
-            object @object = null;
-
-            if (objectWrapper.Value is IGH_GeometricGoo)
+            if (targetType == typeof(ISegmentable3D) || typeof(ISegmentable3D).IsAssignableFrom(targetType))
```
Comment (SAM-BIM-AI) [inline Grasshopper/SAM.Geometry.Grasshopper/Query/TryGetSAMGeometries.cs]: **Ordering bug: this block shadows the `Polyline3D` / `Polygon3D` / `Segment3D` blocks below.**

`Segment3D`, `Polyline3D`, and `Polygon3D` all implement `ISegmentable3D`, so `typeof(ISegmentable3D).IsAssignableFrom(targetType)` is `true` for those `T` — and this block runs first, so their dedicated blocks (lines 350/420/456) are never reached.

Concrete failure: `T = Polyline3D` with a `Face3D` input falls into `if (value is Face3D)` here and executes `result.Add((T)(object)segmentable3D)` where `segmentable3D` is the edge (a `Polygon3D`). `(Polyline3D)(object)polygon3D` throws `InvalidCastException` at runtime. `T = Segment3D` with a `Polycurve3D` input hits the same via line 282 (`(Segment3D)(object)polyline3D`).

Since the only interface `T` callers actually request is `ISegmentable3D` itself, the safe fix is an exact match here:
```csharp
if (targetType == typeof(ISegmentable3D))
```
Alternatively, move the concrete-type blocks (`Segment3D`, `Polyline3D`, `Polygon3D`) above this interface block so most-specific wins.

File: Grasshopper/SAM.Geometry.Grasshopper/Query/TryGetSAMGeometries.cs [hunk 2]
```diff
@@ -6,314 +6,603 @@
...
+                    if (face3DObject.Face3D.GetExternalEdge3D() is Polygon3D extPolygon)
                     {
-                        if (Polycurve3D.TryGetPolyline3D((Polycurve3D)@object, out Polyline3D polyline3D) && polyline3D != null)
+                        result.Add((T)(object)extPolygon);
+                        return;
+                    }
+                }
+            }
+
+            if (targetType == typeof(Segment3D) || typeof(Segment3D).IsAssignableFrom(targetType))
```
Comment (SAM-BIM-AI) [inline Grasshopper/SAM.Geometry.Grasshopper/Query/TryGetSAMGeometries.cs]: This `Segment3D` block (and the `Polyline3D`/`Polygon3D` blocks above) is currently unreachable for `Face3D`/`Polycurve3D` inputs because the `ISegmentable3D` block at line 276 matches `T = Segment3D` first via `IsAssignableFrom` and `return`s. Fixing the dispatch there (exact equality or reordering) is what makes this code path live.

File: Grasshopper/SAM.Geometry.Grasshopper/Query/TryGetSAMGeometries.cs [hunk 3]
```diff
@@ -6,314 +6,603 @@
...
+            {
+                return ghRectangle.ToSAM();
+            }
+
+            if (geometricGoo is GH_Circle ghCircle)
+            {
+                return ghCircle.ToSAM();
+            }
+
+            return geometricGoo.ToSAM(true);
```
Comment (SAM-BIM-AI) [inline Grasshopper/SAM.Geometry.Grasshopper/Query/TryGetSAMGeometries.cs:233]: Nit: this generic `geometricGoo.ToSAM(true)` fallback is only reached for `IGH_GeometricGoo` types not matched above. Worth a quick confirmation that every input type the PR summary lists (`GH_Vector`, `GH_Plane`, etc.) is in fact an `IGH_GeometricGoo` — any that are only `IGH_Goo` would skip `ConvertGeometricGooToSAM` entirely and fall through to the `ScriptVariable()` path instead. A test per goo type would lock this down.

**Review by SAM-BIM-AI — COMMENTED** [review]
## Review — `TryGetSAMGeometries` refactor

Nice cleanup overall: killing the `as dynamic` DLR dispatch, extracting a single recursive `ProcessValue`, unifying the single/collection entry points, and adding XML docs are all clear wins. The two control-flow fixes (removing the mid-loop `return true;`/`return false;` that abandoned remaining faces/items) are genuine bug fixes and the recursion now handles nested enumerables that the old flat cast missed.

A few things before merge — one is a latent correctness bug, and the main gap is **test coverage** (0 tests added for a 480-line rewrite of a conversion function with many branches).

### Correctness
- **Type-dispatch ordering hazard** (inline on the `ISegmentable3D` block): using `typeof(ISegmentable3D).IsAssignableFrom(targetType)` makes the `ISegmentable3D` block capture `T = Polyline3D`, `Polygon3D`, and `Segment3D` too (all three implement `ISegmentable3D`), because it runs *before* their dedicated blocks. For those `T` with a `Face3D`/`Polycurve3D` input this adds `(T)(object)edge` where the edge is a different concrete type → `InvalidCastException`. No in-repo caller requests those three types today, so it's latent, but the PR deliberately added those blocks, so they should actually work.

### Tests (please add)
This function is pure and highly branchy — ideal for unit tests. Suggested matrix (per target `T`, one direct + one conversion + one collection case):
- `Face3D` from: `Shell`, `Mesh3D`, closed `Polyline3D`, `Polygon3D`/`IClosedPlanar3D`, `IFace3DObject`.
- `ISegmentable3D` / `Plane` / `Point3D` / `Shell` from their respective sources.
- **Regression guards for the two fixed bugs**: a collection where the *first* item doesn't convert but later items do (old code aborted early), and a multi-face `Shell`/collection where edges of faces beyond the first must all be returned.
- Null/empty guards: null wrapper, null `.Value`, empty enumerable.

Happy to help scaffold these against `SAM.Geometry.Tests` if useful.

### Minor / nits
- `ProcessSAMGeometryConversion` calls `IsAssignableFrom` on every invocation — fine for GH interactivity, but the exact-equality approach suggested inline also removes the reflection cost.
- PR body says "Tests executed cleanly" but no test files are in the diff — worth clarifying that means the existing suite still passes, not that new coverage was added.

**Comment by michaldengusiak** [comment]
## Review — `TryGetSAMGeometries` refactor (current state)

Re-reviewed against the latest commit (`abcd6178 Comments implemented`). The core refactor is solid and the intent is well executed:

- ✅ **`InvalidCastException` hazard from the earlier review is resolved.** Concrete target blocks (`Face3D` … `Point3D`, 1–7) now run **before** the interface blocks (`ISegmentable3D`, `IClosedPlanar3D`, 8–9), and *every* add goes through a safe `is T` pattern match. A runtime cast exception is no longer reachable. Confirmed `IClosedPlanar3D` does **not** extend `ISegmentable3D`, so block 8 does not shadow block 9.
- ✅ Killing the `as dynamic` DLR dispatch, the single recursive `ProcessValue`, and unifying the two entry points are clear wins.
- ✅ The two control-flow fixes are genuine bug fixes (mid-loop `return false;` in the `ISegmentable3D` branch and the `return true;` after the first face's edges — both abandoned remaining items/faces).
- ✅ No SAM geometry type implements `IEnumerable`, so the `IEnumerable` branch in `ProcessValue` cannot accidentally hijack a real geometry's conversion.

A few things to confirm/address before merge:

### 1. Behavioral change — non-solid `Brep` → `Shell` target (please confirm intended)
`Convert.ToSAM(Brep)` / `GH_Brep.ToSAM(true)` return `List<ISAMGeometry3D>`: a **single `Shell`** when `brep.IsSolid`, but a **list of `Face3D`** otherwise. With the new enumerate → block-5 path, a **non-solid** brep requested as `List<Shell>` now yields **one single-face `Shell` per face**, whereas the old code returned `false` (nothing). This path is hit by the most common callers (`SAMGeometryShellUnion`, `ShellIntersection`, `ShellDifference`, `SAMAnalyticalAddSpacesByBrep`, …). If shells from open breps are wanted at all, one `Shell` containing all faces is almost certainly more correct than N degenerate 1-face shells. Worth confirming this is deliberate.

### 2. Test coverage (main gap)
A ~430-line rewrite of branchy conversion logic ships with **0 tests**. This function is pure and ideal for unit tests. Suggested matrix (per real target `T`: direct + conversion + collection):
- `Face3D` from `Shell`, `Mesh3D`, closed `Polyline3D`, `Polygon3D`, `IFace3DObject`, solid & non-solid `Brep`.
- `Shell` from solid `Brep` (one shell) vs non-solid `Brep` (see #1) and `Face3D`.
- `Plane` / `Point3D` / `ISegmentable3D` / `ISAMGeometry3D` from their sources (all four have real callers).
- **Regression guards for the two fixed bugs**: a collection whose *first* item doesn't convert but later ones do; a multi-face `Shell`/collection where edges of faces beyond the first must all be returned.
- Null/empty guards: null wrapper, null `.Value`, empty enumerable.

### 3. Duplication / maintainability (nit)
The "extract edges from a `Face3D` / `IFace3DObject`" logic is copy-pasted across blocks 2, 4, 8, and 9. A small `IEnumerable<IClosedPlanar3D> GetEdges(object value)` helper (handling both `Face3D` and `IFace3DObject`) would remove most of the repetition and keep the blocks in sync.

### 4. Minor
- `targetType == typeof(X) || typeof(X).IsAssignableFrom(targetType)` — the first disjunct is redundant (`IsAssignableFrom` already returns true for equal types).
- The `IEnumerable` branch is placed before `ProcessSAMGeometryConversion`; it's correct today only because no SAM geometry implements `IEnumerable`. A one-line comment stating that assumption would protect a future geometry type from silently being iterated instead of converted.
- `ProcessValue`'s `IGH_GeometricGoo` branch `return`s unconditionally even when conversion yields `null`, silently dropping the input. Probably fine, but intentional?

Net: the logic looks correct for all in-repo callers and the two bug fixes are real — I'd just want **#1 confirmed** and ideally **#2 (a few regression tests)** before merging.

**Comment by michaldengusiak** [comment]
## Re-review — `01f026ff "Code reorganized and test added"` ✅

Thanks for the quick turnaround — this addresses every point from the previous review. **LGTM / ready to merge from my side.**

- **#1 (non-solid Brep → Shell):** resolved. `ProcessValue` now aggregates a face collection into a **single** `Shell` before the generic `IEnumerable` branch. Confirmed the solid-brep path is unaffected (the `[Shell]` item falls through to normal handling), and a non-solid brep's `[Face3D…]` now yields one shell rather than N single-face shells.
- **#2 (tests):** resolved. 8 xUnit tests added, including the two regression guards for the fixed control-flow bugs (`CollectionWithFirstInvalidElement_DoesNotAbortEarly`, multi-face `Shell` aggregation). Test project now references `SAM.Geometry.Grasshopper`.
- **#3 (duplication):** resolved via the new `GetEdges(object)` helper, reused across the Polyline3D/Segment3D/ISegmentable3D blocks.
- **#4 (nits):** all addressed — simplified `IsAssignableFrom` guards, the `IEnumerable`-assumption comment, and the `IGH_GeometricGoo` branch now only `return`s on a successful conversion (so failed conversions correctly fall through to the Rhino/`ScriptVariable` fallbacks).

Non-blocking notes (fine to merge as-is):
- The Rhino/GH conversion helpers still aren't unit-tested (needs the Rhino runtime) — tests cover the pure SAM-geometry paths only.
- Cosmetic edge case: an enumerable mixing a `Shell` *and* loose `Face3D`s with target `Shell` would aggregate the loose faces and drop the `Shell`. Not reachable from any current conversion path.

**Review by michaldengusiak — APPROVED** [review]
